lib/ocpp/src/core.c

```diff
@@ -0,0 +1,510 @@
+				      K_SECONDS(keyval->ival));
+		}
+	}
+
+	return ret;
+}
+
+int ocpp_stop_transaction(ocpp_session_handle_t hndl,
+			  int Wh,
+			  uint32_t timeout_ms)
+{
+	ocpp_session_t *sh = (ocpp_session_t *)hndl;
+	ocpp_info_t *ctx;
+	ocpp_upstream_info_t *ui;
+	int uid;
+	char buf[350]; // fix me stack
+	char utc[CISTR25] = {0};
+	ocpp_wamp_rpc_msg_t rmsg = {0};
+	ocpp_msg_fp_t fn;
+	int ret;
+ 
+	if (!ocpp_session_is_valid(sh)) {
+		return -EINVAL;
+	}
+
+	ctx = sh->ctx;
+	if (ctx->state < CP_STATE_READY) {
+		return -EAGAIN;
+	}
+
+	sh->is_active = false;
+	if (atomic_dec(&ctx->mtr_timer_ref_cnt) <= 1) {
+		k_timer_stop(&ctx->mtr_timer);
+	}
+
+	// server offline, accept stop txn and save it in Q.
+	if (ctx->is_cs_offline) {
+	}
+
+	fn = ctx->cfn[StopTransaction];
+	ocpp_get_utc_now(utc);
+	sh->uid = fn(buf, sizeof(buf), sh, Wh, NULL, utc);
+
+	// on response check idtag & update cache !!
+	ui = ctx->ui;
+	rmsg.ctx = ctx;
+	rmsg.msg = buf;
+	rmsg.msg_len = strlen(buf) + 1;
+	rmsg.sndlock = &ui->ws_sndlock;
+	rmsg.rspsig = &ui->ws_rspsig;
+	ret = ocpp_send_to_server(&rmsg, K_MSEC(timeout_ms));
+	return ret;
+}
+
+int ocpp_remote_start_transaction(ocpp_info_t *ctx, internal_msg_t *msg,
+				  char *uid)
+{
+	char buf[350]; // fix me stack
+	char *resp = "Rejected";
+	ocpp_wamp_rpc_msg_t rmsg = {0};
+	ocpp_msg_fp_t fn;
+	int ret;
+ 
+	ret = k_msgq_put(ctx->msgq, msg, K_MSEC(100));
+	if (!ret) {
+		resp = "Accepted";
+	}
+
+	fn = ctx->cfn[RemoteStartTransaction];
+	fn(buf, sizeof(buf), resp, uid);
+
+	rmsg.ctx = ctx;
+	rmsg.msg = buf;
+	rmsg.msg_len = strlen(buf) + 1;
+	ret = ocpp_send_to_server(&rmsg, K_SECONDS(2));
+
+	return ret;
+}
+
+int ocpp_remote_stop_transaction(ocpp_info_t *ctx, internal_msg_t *msg,
+				 int idtxn, char *uid)
+{
+	char buf[350]; // fix me stack
+	char *resp = "Rejected";
+	ocpp_wamp_rpc_msg_t rmsg = {0};
+	ocpp_msg_fp_t fn;
+	sys_snode_t *node;
+	ocpp_session_t *sh;
+	bool is_found = false;
+	int ret;
+
+	k_mutex_lock(&ctx->ilock, K_FOREVER);
+	SYS_SLIST_FOR_EACH_NODE(&ctx->slist, node) {
+		sh = to_session(node);
+		if (sh->is_active &&
+		    sh->idtxn == idtxn) {
+			is_found = true;
+			break;
+		}
+	}
+	k_mutex_unlock(&ctx->ilock);
+
+	if (is_found) {
+		msg->usr.stop_charge.id_con = sh->idcon;
+		ret = k_msgq_put(ctx->msgq, msg, K_MSEC(100));
+		if (!ret) {
+			resp = "Accepted";
+		}
+	}
+
+	fn = ctx->cfn[RemoteStopTransaction];
+	fn(buf, sizeof(buf), resp, uid);
+
+	rmsg.ctx = ctx;
+	rmsg.msg = buf;
+	rmsg.msg_len = strlen(buf) + 1;
+	ret = ocpp_send_to_server(&rmsg, K_SECONDS(2));
+
+	return ret;
+}
+
+int ocpp_unlock_connector(ocpp_info_t *ctx, internal_msg_t *msg, char *uid)
+{
+	char buf[350]; // fix me stack
+	char *resp = "UnlockFailed";
+	ocpp_wamp_rpc_msg_t rmsg = {0};
+	ocpp_msg_fp_t fn;
+	sys_snode_t *node;
+	bool is_found = false;
+	int ret;
+
+	ret = k_msgq_put(ctx->msgq, msg, K_MSEC(100));
+	if (!ret) {
+		resp = "Unlocked";
+	}
+
+	fn = ctx->cfn[UnlockConnector];
+	fn(buf, sizeof(buf), resp, uid);
```

Comment:
Using a fn variable seems pointless just for this one call, why not just do `ctx->cfn[UnlockConnector](buf, ....);`

Reply:
made for better code readability